libraries/AP_GPS/AP_GPS_NMEA.cpp

```diff
@@ -316,8 +319,33 @@ bool AP_GPS_NMEA::_term_complete()
                     state.location.lng  = _new_longitude;
                     state.num_sats      = _new_satellite_count;
                     state.hdop          = _new_hdop;
-                    // To-Do: add support for proper reporting of 2D and 3D fix
-                    state.status        = AP_GPS::GPS_OK_FIX_3D;
+                    switch(_new_quality_indicator) {
+                    case 0: // Fix not available or invalid
+                        state.status = AP_GPS::NO_FIX;
+                        break;
+                    case 1: // GPS SPS Mode, fix valid
+                        state.status = AP_GPS::GPS_OK_FIX_3D;
+                        break;
+                    case 2: // Differential GPS, SPS Mode, fix valid
+                        state.status = AP_GPS::GPS_OK_FIX_3D_DGPS;
+                        break;
+                    case 3: // GPS PPS Mode, fix valid
+                        state.status = AP_GPS::GPS_OK_FIX_3D;
+                        break;
+                    case 4: // Real Time Kinematic. System used in RTK mode with fixed integers
+                        state.status = AP_GPS::GPS_OK_FIX_3D_RTK_FIXED;
+                        break;
+                    case 5: // Float RTK. Satellite system used in RTK mode, floating integers
+                        state.status = AP_GPS::GPS_OK_FIX_3D_RTK_FLOAT;
+                        break;
+                    case 6: // Estimated (dead reckoning) Mode
+                        state.status = AP_GPS::GPS_OK_FIX_3D;
```

Comment:
Dead reckoning should not be accepted as a GPS fix type, and the EKF should be the only source of dead reckoning.